homeassistant/components/tankerkoenig/__init__.py

```diff
@@ -69,87 +79,84 @@ async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
+        _async_update_listener
     )
 
+    hass.data.setdefault(DOMAIN, {})
+    hass.data[DOMAIN][entry.unique_id] = tankerkoenig
+
+    hass.config_entries.async_setup_platforms(entry, [Platform.SENSOR])
+
     return True
 
 
-class TankerkoenigData:
-    """Get the latest data from the API."""
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Unload Tankerkoenig config entry."""
+    unload_ok = await hass.config_entries.async_unload_platforms(
+        entry, [Platform.SENSOR]
+    )
+    if unload_ok:
+        entry_data: TankerkoenigData = hass.data[DOMAIN][entry.unique_id]
+        assert entry_data.undo_update_listener is not None
+        entry_data.undo_update_listener()
+        hass.data[DOMAIN].pop(entry.unique_id)
 
-    def __init__(self, hass, conf):
-        """Initialize the data object."""
-        self._api_key = conf[CONF_API_KEY]
-        self.stations = {}
-        self.fuel_types = conf[CONF_FUEL_TYPES]
-        self.update_interval = conf[CONF_SCAN_INTERVAL]
-        self.show_on_map = conf[CONF_SHOW_ON_MAP]
-        self._hass = hass
+    return unload_ok
 
-    def setup(self, latitude, longitude, radius, additional_stations):
-        """Set up the tankerkoenig API.
 
-        Read the initial data from the server, to initialize the list of fuel stations to monitor.
-        """
-        _LOGGER.debug("Fetching data for (%s, %s) rad: %s", latitude, longitude, radius)
-        try:
-            data = pytankerkoenig.getNearbyStations(
-                self._api_key, latitude, longitude, radius, "all", "dist"
-            )
-        except pytankerkoenig.customException as err:
-            data = {"ok": False, "message": err, "exception": True}
-        _LOGGER.debug("Received data: %s", data)
-        if not data["ok"]:
-            _LOGGER.error(
-                "Setup for sensors was unsuccessful. Error occurred while fetching data from tankerkoenig.de: %s",
-                data["message"],
-            )
-            return False
+async def _async_update_listener(hass: HomeAssistant, entry: ConfigEntry) -> None:
+    """Handle options update."""
+    await hass.config_entries.async_reload(entry.entry_id)
 
-        # Add stations found via location + radius
-        if not (nearby_stations := data["stations"]):
-            if not additional_stations:
-                _LOGGER.error(
-                    "Could not find any station in range."
-                    "Try with a bigger radius or manually specify stations in additional_stations"
-                )
-                return False
-            _LOGGER.warning(
-                "Could not find any station in range. Will only use manually specified stations"
-            )
-        else:
-            for station in nearby_stations:
-                self.add_station(station)
 
-        # Add manually specified additional stations
-        for station_id in additional_stations:
+class TankerkoenigData:
```

Comment:
Why not a data update coordinator?

Reply:
yeah ... it is existing code 🙈 sure I could rework this to be a update coordinator, but would like to do this in a later PR?

Reply:
as there is nothing but trash on TV tonight, I found time for this - see 5464e3f17bb7edbd1acdde2428f3ba77abef3b8e